### ISSUE TITLE
`python-version` based syntax checks should not block

### ISSUE THREAD

**Issue by A5rocks** [python/mypy#21178]
**Feature**

Things like:
```
Type statement is only supported in Python 3.12 and greater
```
should not block running mypy if running on Python 3.12+.

**Pitch**

Blocking errors suck!

**Comment by Mkashiii** [comment]
I can help LEts Connect 
M.kashiact@gmail.com

**Comment by BHUVANSH855** [comment]
Hi! I'm working on this issue and have reproduced it with a regression test.

While investigating, I found that `fastparse` returns an empty `MypyFile` after a `SyntaxError`, so simply making these errors non-blocking doesn't allow semantic analysis to continue.

Is the intended approach to recover by reparsing with the runtime parser (when running on a newer Python version), or is there another preferred direction for `fastparse`?

I'd like to align the implementation with the intended design before opening a PR.

**Comment by A5rocks** [comment]
These shouldn't raise a SyntaxError? This is using 3.13 to parse 3.12 code, but with `--python-version 3.11`. I don't think it's possible to make parsing 3.12 code with 3.11 non-blocking.

If you mean that fastparse bundles a 3.11 parser and uses it, TIL! I guess I don't know if this is possible then.

**Comment by BHUVANSH855** [comment]
Thanks! I think I found where the confusion is.

I'm reproducing this while running mypy on Python 3.12, not Python 3.11.

`fastparse` calls:

```python
ast.parse(..., feature_version=options.python_version[1])
```

so when the target is `--python-version 3.11`, CPython intentionally parses the source as Python 3.11 and raises errors like:

```
Type statement is only supported in Python 3.12 and greater
```

I also checked the history: this behavior appears to come from the original switch to `ast.parse(feature_version=...)` (#6539), and I couldn't find any existing tests or comments indicating that these version-based `SyntaxError`s were intentionally meant to be blocking.

Given that, is the expectation that `fastparse` should continue using `feature_version` this way, or should these version-gated syntax errors be handled differently?

**Comment by A5rocks** [comment]
My opinion is that `fastparse` should make these non-blocking, but I didn't realize `ast.parse` supported a `feature_version=...` parameter! I guess this might be more contentious than I expected.

**Comment by A5rocks** [comment]
OK so I'm not convinced this is possible to solve as is due to things already mentioned here. However, the reason I even care about this is because I'm installing dependencies conditional on python version, so pull in one that uses new type features, which means that I can't run typechecking locally due to having a new Python installed.

This particular root cause can be addressed and I threw together a PR. If someone runs into this issue w/o that as a root cause, IMO they should make a new issue.
